### PR TITLE
d/aws_prefix_list: add support for managed prefix lists

### PR DESCRIPTION
**Pull request by roberth-k**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #14068, #14109 (both must be merged before this change), #13986

Ignore changes related to the `aws_prefix_list` and `aws_prefix_list_entry` resources: code from #14068 are necessary in order to be able to test changes to the data source.

As stated in the discussion of #14068:

> The DescribeManagedPrefixLists API is a superset of DescribePrefixLists, even returning AWS's own prefix lists by default. A separate `aws_prefix_list` resource would have to artificially insert a filter by `owner-id` in order to produce only customer-managed prefix lists. It seemed appropriate to present an unambiguous data source, though I did at one point consider aliasing the two names.

However, I'm happy to expose managed prefix lists as a separate data source if that's the maintainers' preference.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
ENHANCEMENTS:
- data-source/aws_prefix_list: Support querying managed prefix lists
- data-source/aws_prefix_list: Add `owner_id`, `address_family`, `arn`, `max_entries`, `tags` attributes
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS=-run=TestA
ccDataSourceAwsPrefixList_
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsPrefixList_ -timeout 120m
--- PASS: TestAccDataSourceAwsPrefixList_nameDoesNotOverrideFilter (4.55s)
--- PASS: TestAccDataSourceAwsPrefixList_matchesTooMany (4.55s)
--- PASS: TestAccDataSourceAwsPrefixList_filter (23.59s)
--- PASS: TestAccDataSourceAwsPrefixList_basic (23.61s)
--- PASS: TestAccDataSourceAwsPrefixList_managedPrefixList (32.93s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       34.474s
```

### CODE REVIEW

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by robcoward** [comment]
Any chance of getting this merged and released ?

**Comment by lmmattr** [comment]
@roberth-k Do you know when this will be merged? Would love to have this in.

**Comment by ewbankkit** [comment]
@lmmattr Support for Managed Prefix Lists is on the [roadmap](https://github.com/terraform-providers/terraform-provider-aws/pull/15898) for the upcoming quarter.

**Comment by bflad** [comment]
(Note: breaking change label applied due to containing commits from other pull request: https://github.com/hashicorp/terraform-provider-aws/pull/14109#pullrequestreview-544276108 -- will remove the label when those feedback items are also addressed in this pull request)

File: aws/data_source_aws_prefix_list.go
Comment (bflad) [inline aws/data_source_aws_prefix_list.go:62]: Just a very forward comment here since it will dramatically change the implementation -- rather than change the API call associated with the existing data source, which can cause issues in environments that utilize restrictive IAM permissions or with AWS "compatible" APIs that may not support the new API call, the preference would be to create a separate data source so operators can choose the appropriate solution. 👍 

It may be easiest to close this pull request and separately submit a pull request with just a new `aws_ec2_managed_prefix_list` data source that for now just handles AWS managed prefix lists. Once #14068 is merged, we can add any remaining functionality to the new data source.

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @roberth-k 👋  Thank you _again_ for all the work so far. 😄  It felt important to drop this first comment in since it might be easiest to submit a net-new PR without the baggage of the prior commits. I can review that implementation immediately if/when its available. 👍

File: aws/data_source_aws_prefix_list.go
Comment (roberth-k) [reply aws/data_source_aws_prefix_list.go:62]: Thanks @bflad,

> may not support the new API call,

I hadn't considered this detail, and it makes the right course clear in retrospect. Indeed, the managed prefix list API-s require a different set of IAM actions.

Just to confirm, is `aws_ec2_managed_prefix_list` (as well as `aws_ec2_managed_prefix_list_entry` in the resource PR) the preferred nomenclature?

Comment (bflad) [reply aws/data_source_aws_prefix_list.go:62]: Yes, please, and thank you regarding the naming! We can separately discuss the merits of an `_entry` resource in the other pull request.

**Comment by roberth-k** [comment]
@bflad Superseded by https://github.com/hashicorp/terraform-provider-aws/pull/16738.

**Pull request closed without merging**
